tests/app/main/views/test_users.py

```diff
@@ -362,6 +362,137 @@ def test_download_csv_for_on_framework_only(self):
+        ("admin-manager", 403),
+    ])
+    def test_supplier_csv_is_only_accessible_to_specific_user_roles(self, role, expected_code):
+        self.user_role = role
+        self.data_api_client.export_suppliers.return_value = {"suppliers": []}
+        self.data_api_client.find_frameworks.return_value = {"frameworks": [self._valid_framework]}
+        self.data_api_client.get_framework.return_value = {"frameworks": self._valid_framework}
+
+        response = self.client.get(
+            '/admin/frameworks/{}/suppliers/download'.format(self._valid_framework['slug'])
+        )
+
+        assert response.status_code == expected_code
+
+    def test_download_supplier_csv_for_framework_users(self):
+        list_of_expected_supplier_results = [
+            {
+                'supplier_id': 1,
+                'application_result': 'no result',
+                'application_status': 'no_application',
+                'declaration_status': 'unstarted',
+                'framework_agreement': False,
+                'supplier_name': "Supplier 1",
+                'supplier_organisation_size': "small",
+                'duns_number': "100000001",
+                'registered_name': 'Registered Supplier Name 1',
+                'companies_house_number': 'ABC123',
+                "published_services_count": {
+                    "digital-outcomes": 0,
+                    "digital-specialists": 0,
+                    "user-research-studios": 0,
+                    "user-research-participants": 0,
+                },
+                "contact_information": {
+                    'contact_name': 'Contact for Supplier 1',
+                    'contact_email': '1@contact.com',
+                    'contact_phone_number': '12345',
+                    'address_first_line': '7 Gem Lane',
+                    'address_city': 'Cantelot',
+                    'address_postcode': 'CN1A 1AA',
+                    'address_country': 'country:GB',
+                },
+                'variations_agreed': '',
+            }
+        ]
+        self.data_api_client.export_suppliers.return_value = {"suppliers": list_of_expected_supplier_results}
+        self.data_api_client.find_frameworks.return_value = {"frameworks": [self._valid_framework]}
+
+        self.data_api_client.get_framework.return_value = {"frameworks": self._valid_framework}
+        response = self.client.get(
+            '/admin/frameworks/{}/suppliers/download'.format(
+                self._valid_framework['slug']
+            )
+        )
+        assert response.status_code == 200
+        assert response.mimetype == 'text/csv'
+        assert (response.headers['Content-Disposition'] ==
+                'attachment;filename=suppliers-on-digital-outcomes-and-specialists-2.csv')
+
+        rows = [line.split(",") for line in response.get_data(as_text=True).splitlines()]
+
+        assert len(rows) == len(list_of_expected_supplier_results) + 1
+        expected_top_level_headings = [
```

Comment:
I think this splitting the `expected_headings` into 3 different variables here mirrors the code a bit too much. This is only done because the code in the view does it as far as I can tell.

We should be testing the code as if we didn't know how it was written. Only test the expected output rather than testing _how_ we get the expected outcome.

Really, technically, what we want to assert is that the headings are correct. Not that the headings are equal to `expected_top_level_headings + expected_service_count_headings + expected_contact_info_headings`. In this case it is the same thing but it doesn't add value to the test to split them because someone reading this test might wonder, 'Why should the headings equal 1 + 2 + 3?' Then they have to look at the code to find out that it's not important.

Reply:
Can you tell I wrote this test in about 5 minutes 😹 I'll have another look this morning and tidy it up.